src/main/java/org/opensearch/agent/tools/RAGTool.java

```diff
@@ -0,0 +1,275 @@
+ */
+@Log4j2
+@Setter
+@Getter
+@ToolAnnotation(RAGTool.TYPE)
+public class RAGTool extends AbstractRetrieverTool {
+    public static final String TYPE = "RAGTool";
+    public static String DEFAULT_DESCRIPTION =
+        "Use this tool to retrieve helpful information to optimize the output of the large language model to answer questions.";
+    public static final String INFERENCE_MODEL_ID_FIELD = "inference_model_id";
+    public static final String EMBEDDING_MODEL_ID_FIELD = "embedding_model_id";
+    public static final String EMBEDDING_FIELD = "embedding_field";
+    public static final String OUTPUT_FIELD = "output_field";
+    private String name = TYPE;
+    private String description = DEFAULT_DESCRIPTION;
+    private Client client;
+    private String inferenceModelId;
+    private NamedXContentRegistry xContentRegistry;
+    private String index;
+    private String embeddingField;
+    private String[] sourceFields;
+    private String embeddingModelId;
+    private Integer docSize;
+    private Integer k;
+    @Setter
+    private Parser inputParser;
+    @Setter
+    private Parser outputParser;
+
+    @Builder
+    public RAGTool(
+        Client client,
+        NamedXContentRegistry xContentRegistry,
+        String index,
+        String embeddingField,
+        String[] sourceFields,
+        Integer k,
+        Integer docSize,
+        String embeddingModelId,
+        String inferenceModelId
+    ) {
+        super(client, xContentRegistry, index, sourceFields, docSize);
+        this.client = client;
+        this.xContentRegistry = xContentRegistry;
+        this.index = index;
+        this.embeddingField = embeddingField;
+        this.sourceFields = sourceFields;
+        this.embeddingModelId = embeddingModelId;
+        this.docSize = docSize == null ? DEFAULT_DOC_SIZE : docSize;
+        this.k = k == null ? DEFAULT_K : k;
+        this.inferenceModelId = inferenceModelId;
+
+        outputParser = new Parser() {
+            @Override
+            public Object parse(Object o) {
+                List<ModelTensors> mlModelOutputs = (List<ModelTensors>) o;
+                return mlModelOutputs.get(0).getMlModelTensors().get(0).getDataAsMap().get("response");
+            }
+        };
+    }
+
+    // getQueryBody is not used in RAGTool
+    @Override
+    protected String getQueryBody(String queryText) {
+        return queryText;
+    }
+
+    @Override
+    public <T> void run(Map<String, String> parameters, ActionListener<T> listener) {
+        String input = null;
+
+        try {
+            if (!this.validate(parameters)) {
+                throw new IllegalArgumentException("[" + INPUT_FIELD + "] is null or empty, can not process it.");
+            }
+            String question = parameters.get(INPUT_FIELD);
+            input = gson.fromJson(question, String.class);
+        } catch (Exception e) {
+            log.error("[" + INPUT_FIELD + "] is null or empty, can not process it.", e);
```

Comment:
Why this error message same with ling 115 ?  Suggest change to a more general error message as the try block is mainly to catch exception from `gson.fromJson`.

Reply:
the exception from gson.fromJson is changed to `new IllegalArgumentException("Failed to read question from " + INPUT_FIELD)`